homeassistant/components/tplink/__init__.py

```diff
@@ -64,152 +59,89 @@
+
+    legacy_entry = None
+    config_entries_by_mac = {}
+    for entry in hass.config_entries.async_entries(DOMAIN):
+        if async_entry_is_legacy(entry):
+            legacy_entry = entry
+        elif entry.unique_id:
+            config_entries_by_mac[entry.unique_id] = entry
 
-    hass.data[DOMAIN] = {}
-    hass.data[DOMAIN][ATTR_CONFIG] = conf
+    domain_data[DISCOVERED_DEVICES] = {
+        dr.format_mac(device.mac): device
+        for device in (await Discover.discover()).values()
+    }
+
+    if legacy_entry:
+        async_migrate_legacy_entries(hass, config_entries_by_mac, legacy_entry)
 
     if conf is not None:
-        hass.async_create_task(
-            hass.config_entries.flow.async_init(
-                DOMAIN, context={"source": config_entries.SOURCE_IMPORT}
-            )
-        )
+        async_migrate_yaml_entries(hass, conf)
+
+    if domain_data[DISCOVERED_DEVICES]:
+        async_trigger_discovery(hass, domain_data[DISCOVERED_DEVICES])
 
     return True
 
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up TPLink from a config entry."""
-    config_data = hass.data[DOMAIN].get(ATTR_CONFIG)
-    if config_data is None and entry.data:
-        config_data = entry.data
-    elif config_data is not None:
-        hass.config_entries.async_update_entry(entry, data=config_data)
-
-    device_registry = dr.async_get(hass)
-    tplink_devices = dr.async_entries_for_config_entry(device_registry, entry.entry_id)
-    device_count = len(tplink_devices)
-    hass_data: dict[str, Any] = hass.data[DOMAIN]
-
-    # These will contain the initialized devices
-    hass_data[CONF_LIGHT] = []
-    hass_data[CONF_SWITCH] = []
-    hass_data[UNAVAILABLE_DEVICES] = []
-    lights: list[SmartDevice] = hass_data[CONF_LIGHT]
-    switches: list[SmartDevice] = hass_data[CONF_SWITCH]
-
-    # Add static devices
-    static_devices = SmartDevices()
-    if config_data is not None:
-        static_devices = await get_static_devices(config_data)
-
-        lights.extend(static_devices.lights)
-        switches.extend(static_devices.switches)
-
-    # Add discovered devices
-    if config_data is None or config_data[CONF_DISCOVERY]:
-        discovered_devices = await async_discover_devices(
-            hass, static_devices, device_count
-        )
+    if async_entry_is_legacy(entry):
+        return True
 
-        lights.extend(discovered_devices.lights)
-        switches.extend(discovered_devices.switches)
+    legacy_entry: ConfigEntry | None = None
+    for config_entry in hass.config_entries.async_entries(DOMAIN):
+        if async_entry_is_legacy(config_entry):
+            legacy_entry = config_entry
+            break
 
-    if lights:
-        _LOGGER.debug(
-            "Got %s lights: %s", len(lights), ", ".join(d.host for d in lights)
-        )
+    if legacy_entry is not None:
+        await async_migrate_entities_devices(hass, legacy_entry.entry_id, entry)
```

Comment:
Why not do this in `async_setup` ?

Reply:
Its a chicken-egg issue. We don't know the config entry id until its created so we can't migrate it in async_setup.